autode/opt/coordinates/primitives.py

```diff
@@ -0,0 +1,222 @@
+import numpy as np
+from abc import ABC, abstractmethod
+from typing import Tuple
+
+
+class Primitive(ABC):
+    """Primitive internal coordinate"""
+
+    @abstractmethod
+    def __call__(self,
+                 x: 'autode.opt.coordinates.CartesianCoordinates'
+                 ) -> float:
+        """Return the value of this PIC given a set of cartesian coordinates"""
+
+    @abstractmethod
+    def derivative(self,
+                   i:         int,
+                   component: str,
+                   x:         'autode.opt.coordinates.CartesianCoordinates'
+                   ) -> float:
+        r"""
+        Calculate the derivative with respect to a cartesian coordinate
+
+        .. math::
+
+            \frac{dq}
+                  {d\boldsymbol{X}_{i, k}} {\Bigg\rvert}_{X=X0}
+
+        where :math:`q` is the primitive coordinate and :math:`\boldsymbol{X}`
+        are the cartesian coordinates.
+
+        -----------------------------------------------------------------------
+        Arguments:
+            i: Cartesian index to take the derivative with respect to;
+                0-N for N atoms
+
+            component: Cartesian component (x, y, z) to take the derivative
+                       with respect to. {'x', 'y', 'z'}
+
+            x: Cartesian coordinates
+
+        Returns:
+            (float): Derivative
+        """
+
+    @staticmethod
+    def _cart_component_to_idx(component: str) -> int:
+        """Convert the name of a Cartesian component (or direction) to an
+        index used for array indexing"""
+
+        if component == 'x':
+            return 0
+
+        if component == 'y':
+            return 1
+
+        if component == 'z':
+            return 2
+
+        else:
+            raise ValueError(f'Cannot convert component {component} to an '
+                             f'index. Must be one of x, y, z')
+
+    @abstractmethod
+    def __eq__(self, other):
+        """Comparison of two primitive coordinates"""
+
+
+class _DistanceFunction(Primitive, ABC):
+    """Function of a distance between two atoms"""
+
+    def __init__(self,
+                 idx_i: int,
+                 idx_j: int):
+        """
+        Function of a distance between a pair of atoms
+
+        .. math::
+
+            q = f(|\boldsymbol{X}_i - \boldsymbol{X}_j|)
+
+        for a set of cartesian coordinates :math:`\boldsymbol{X}`.
+        -----------------------------------------------------------------------
+        Arguments:
+            idx_i: Atom index
+
+            idx_j: Atom index
+        """
+
+        self.idx_i = int(idx_i)
+        self.idx_j = int(idx_j)
+
+    def __eq__(self, other) -> bool:
+        """Equality of two distance functions"""
+
+        return (isinstance(other, self.__class__)
+                and other._ordered_idxs == self._ordered_idxs)
```

Comment:
^ Calling `isinstance` in a `__eq__` definition looks weird to me. If `other` isn't a DistanceFunction object, then I would expect an exception rather than `False`

Reply:
Seemingly it's more pythonic to not throw anything here e.g.

```python
>>> 1 == 1.0
True
>>> 1 == 'a string'
False
```

Reply:
Aah okay, you're right. This always gets me :)